File: ch4/01_dfs.java
import java.util.Collections;
import java.util.Iterator;
import java.util.Scanner;
import java.util.Stack;
import java.util.Vector;
import java.io.*;

public class ch4_01_dfs {
  private static final int DFS_WHITE = -1; // normal DFS
  private static final int DFS_BLACK = 1;
  private static final int DFS_GRAY = 2;
  private static Vector < Vector < IntegerPair > > AdjList = 
             new Vector < Vector < IntegerPair > >();
  private static Vector < Integer > dfs_num, dfs_low, dfs_parent;
  private static Vector < Boolean > articulation_vertex, visited;
  private static Stack < Integer > S; // additional information for SCC
  private static Vector < Integer > topologicalSort; // additional information for toposort
  private static int numComp, dfsNumberCounter, dfsRoot, rootChildren;

  private static void initDFS(int V) { // used in normal DFS
    dfs_num = new Vector < Integer > ();
    dfs_num.addAll(Collections.nCopies(V, DFS_WHITE));
    numComp = 0;
  }

  private static void initGraphCheck(int V) {
    initDFS(V);
    dfs_parent = new Vector < Integer > ();
    dfs_parent.addAll(Collections.nCopies(V, 0));
    numComp = 0;
  }

  private static void initArticulationPointBridge(int V) {
    initGraphCheck(V);
    dfs_low = new Vector < Integer > ();
    dfs_low.addAll(Collections.nCopies(V, 0));
    articulation_vertex = new Vector < Boolean > ();
    articulation_vertex.addAll(Collections.nCopies(V, false));
    dfsNumberCounter = 0;
  }

  private static void initTarjanSCC(int V) {
    initGraphCheck(V);
    dfs_low = new Vector < Integer > ();
    dfs_low.addAll(Collections.nCopies(V, 0));
    dfsNumberCounter = 0;
    numComp = 0;
    S = new Stack < Integer > ();
    visited = new Vector < Boolean > ();
    visited.addAll(Collections.nCopies(V, false));
  }

  private static void initTopologicalSort(int V) {
    initDFS(V);
    topologicalSort = new Vector < Integer > ();
  }

  private static void dfs(int u) { // DFS for normal usage
    System.out.printf(" %d", u); // this vertex is visited
    dfs_num.set(u, DFS_BLACK); // mark as visited
    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) { // try all neighbors v of vertex u
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE) // avoid cycle
        dfs(v.first()); // v is a (neighbor, weight) pair
    }
  }

  private static void floodfill(int u, int color) {
    dfs_num.set(u, color); // not just a generic DFS_BLACK
    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) { // try all neighbors v of vertex u
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE) // avoid cycle
        floodfill(v.first(), color); // v is a (edge, weight) pair
    }
  }

  private static void graphCheck(int u) { // DFS for checking graph edge properties...
    dfs_num.set(u, DFS_GRAY); // color this as DFS_GRAY (temporary)
    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) { // traverse this AdjList
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE) { // DFS_GRAY to DFS_WHITE
        // System.out.printf("  Tree Edge (%d, %d)\n", u, v.first());
        dfs_parent.set(v.first(), u); // parent of this children is me
        graphCheck(v.first());
      }
      else if (dfs_num.get(v.first()) == DFS_GRAY) { // DFS_GRAY to DFS_GRAY
        if (v.first() == dfs_parent.get(u))
          System.out.printf(" Bidirectional Edge (%d, %d) - (%d, %d)\n", u, v.first(), v.first(), u);
        else
          System.out.printf(" Back Edge (%d, %d) (Cycle)\n", u, v.first());
      }
      else if (dfs_num.get(v.first()) == DFS_BLACK) // DFS_GRAY to DFS_BLACK
        System.out.printf(" Forward/Cross Edge (%d, %d)\n", u, v.first());
    }
    dfs_num.set(u, DFS_BLACK); // now color this as DFS_BLACK (DONE)
  }

  private static void articulationPointAndBridge(int u) {
    dfs_low.set(u, dfsNumberCounter);
    dfs_num.set(u, dfsNumberCounter++); // dfs_low[u] <= dfs_num[u]
    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) { // try all neighbors v of vertex u
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE) { // a tree edge
        dfs_parent.set(v.first(), u); // parent of this children is me
        if (u == dfsRoot) // special case
          rootChildren++; // count children of root
        articulationPointAndBridge(v.first());
        if (dfs_low.get(v.first()) >= dfs_num.get(u)) // for articulation point
          articulation_vertex.set(u, true); // store this information first
        if (dfs_low.get(v.first()) > dfs_num.get(u)) // for bridge
          System.out.printf(" Edge (%d, %d) is a bridge\n", u, v.first());
        dfs_low.set(u, Math.min(dfs_low.get(u), dfs_low.get(v.first()))); // update dfs_low[u]
      }
      else if (v.first() != dfs_parent.get(u)) // a back edge and not direct cycle
        dfs_low.set(u, Math.min(dfs_low.get(u), dfs_num.get(v.first()))); // update dfs_low[u]
    }
  }

  private static void tarjanSCC(int u) {
    dfs_num.set(u, dfsNumberCounter);
    dfs_low.set(u, dfsNumberCounter++); // dfs_low[u] <= dfs_num[u]
    S.push(u); // store u according to order of visitation
    visited.set(u, true);

    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) { // try all neighbors v of vertex u
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE) // a tree edge
        tarjanSCC(v.first());
      if (visited.get(v.first())) // condition for update
        dfs_low.set(u, Math.min(dfs_low.get(u), dfs_low.get(v.first())));
    }

    if (dfs_low.get(u) == dfs_num.get(u)) { // if this is a root (start) of an SCC
      System.out.printf("SCC %d: ", ++numComp);
      while (true) {
        int v = S.peek(); S.pop(); visited.set(v, false);
        System.out.printf(" %d", v);
        if (u == v) break;
      }
      System.out.printf("\n");
    }
  }

  private static void topoVisit(int u) {
    dfs_num.set(u, DFS_BLACK);
    Iterator it = AdjList.get(u).iterator();
    while (it.hasNext()) {
      IntegerPair v = (IntegerPair)it.next();
      if (dfs_num.get(v.first()) == DFS_WHITE)
        topoVisit(v.first());
    }
    topologicalSort.add(u);
  }

  private static void printThis(String message) {
    System.out.printf("==================================\n");
    System.out.printf("%s\n", message);
    System.out.printf("==================================\n");
  }

  public static void main(String[] args) throws Exception {
    int i, j, V, total_neighbors, id, weight;

    /*
    // Use the following input:
    // Graph in Figure 4.1
    9
    1 1 0
    3 0 0 2 0 3 0
    2 1 0 3 0
    3 1 0 2 0 4 0
    1 3 0
    0
    2 7 0 8 0
    1 6 0
    1 6 0

    // Example of directed acyclic graph in Figure 4.4 (for toposort)
    8
    2 1 0 2 0
    2 2 0 3 0
    2 3 0 5 0
    1 4 0
    0
    0
    0
    1 6 0

    // Example of directed graph with back edges
    3
    1 1 0
    1 2 0
    1 0 0

    // Left graph in Figure 4.6/4.7/4.8
    6
    1 1 0
    3 0 0 2 0 4 0
    1 1 0
    1 4 0
    3 1 0 3 0 5 0
    1 4 0

    // Right graph in Figure 4.6/4.7/4.8
    6
    1 1 0
    5 0 0 2 0 3 0 4 0 5 0
    1 1 0
    1 1 0
    2 1 0 5 0
    2 1 0 4 0

    // Directed graph in Figure 4.9
    8
    1 1 0
    1 3 0
    1 1 0
    2 2 0 4 0
    1 5 0
    1 7 0
    1 4 0
    1 6 0
    */

    File f = new File("in_01.txt");
    Scanner sc = new Scanner(f);

    V = sc.nextInt();
    AdjList.clear();
    for (i = 0; i < V; i++) {
      Vector < IntegerPair > Neighbor = new Vector < IntegerPair >(); // create vector of pair<int, int> 
      AdjList.add(Neighbor); // store blank vector first
    }

    for (i = 0; i < V; i++) {
      total_neighbors = sc.nextInt();
      for (j = 0; j < total_neighbors; j++) {
        id = sc.nextInt();
        weight = sc.nextInt();
        AdjList.get(i).add(new IntegerPair(id, weight));
      }
    }

    initDFS(V); // call this first before running DFS
    printThis("Standard DFS Demo (the input graph must be UNDIRECTED)");
    for (i = 0; i < V; i++) // for each vertex i in [0..V-1]
      if (dfs_num.get(i) == DFS_WHITE) { // if not visited yet
        System.out.printf("Component %d, visit:", ++numComp);
        dfs(i);
        System.out.printf("\n");
      }
    System.out.printf("There are %d connected components\n", numComp);

    initDFS(V); // call this first before running DFS
    printThis("Flood Fill Demo (the input graph must be UNDIRECTED)");
    for (i = 0; i < V; i++) // for each vertex i in [0..V-1], note that we use our REP macro again
      if (dfs_num.get(i) == DFS_WHITE) // if not visited yet
        floodfill(i, ++numComp);
    for (i = 0; i < V; i++)
      System.out.printf("Vertex %d has color %d\n", i, dfs_num.get(i));

    // make sure that the given graph is DAG
    initTopologicalSort(V);
    printThis("Topological Sort (the input graph must be DAG)");
    for (i = 0; i < V; i++)
      if (dfs_num.get(i) == DFS_WHITE)
        topoVisit(i);
    for (i = topologicalSort.size() - 1; i >= 0; i--) // access from back to front
      System.out.printf(" %d", topologicalSort.get(i));
    System.out.printf("\n");

    initGraphCheck(V);
    printThis("Graph Edges Property Check");
    for (i = 0; i < V; i++)
      if (dfs_num.get(i) == DFS_WHITE) {
        System.out.printf("Component %d:\n", ++numComp);
        graphCheck(i);
      }

    initArticulationPointBridge(V);
    printThis("Articulation Points & Bridges (the input graph must be UNDIRECTED)");
    System.out.printf("Bridges:\n");
    for (i = 0; i < V; i++)
      if (dfs_num.get(i) == DFS_WHITE) {
        dfsRoot = i; rootChildren = 0;
        articulationPointAndBridge(i);
        articulation_vertex.set(dfsRoot, (rootChildren > 1)); // special case
      }

    System.out.printf("Articulation Points:\n");
    for (i = 0; i < V; i++)
      if (articulation_vertex.get(i))
        System.out.printf(" Vertex %d\n", i);

    initTarjanSCC(V);
    printThis("Strongly Connected Components (the input graph must be DIRECTED)");
    for (i = 0; i < V; i++)
      if (dfs_num.get(i) == DFS_WHITE)
        tarjanSCC(i);
  }
}

File: ch4/01_dfs.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> ii;      // In this chapter, we will frequently use these
typedef vector<ii> vii;      // three data type shortcuts. They may look cryptic
typedef vector<int> vi;   // but shortcuts are useful in competitive programming

#define DFS_WHITE -1 // normal DFS, do not change this with other values (other than 0), because we usually use memset with conjunction with DFS_WHITE
#define DFS_BLACK 1

vector<vii> AL;

void printThis(string message) {
  printf("==================================\n");
  printf("%s\n", message.c_str());
  printf("==================================\n");
}

vi dfs_num;     // this variable has to be global, we cannot put it in recursion
int numCC;

void dfs(int u) {     // DFS for normal usage: as graph traversal algorithm
  printf(" %d", u);                               // this vertex is visited
  dfs_num[u] = DFS_BLACK;      // important: we mark this vertex as visited
  for (auto &v : AL[u])              // v is a (neighbor, weight) pair
    if (dfs_num[v.first] == DFS_WHITE)    // important check to avoid cycle
      dfs(v.first);   // recursively visits unvisited neighbors of vertex u
}    // for simple graph traversal, we ignore the weight stored at v.second

// note: this is not the version on implicit graph
void floodfill(int u, int color) {
  dfs_num[u] = color;                            // not just a generic DFS_BLACK
  for (auto &v : AL[u])
    if (dfs_num[v.first] == DFS_WHITE)
      floodfill(v.first, color);
}

vi topoSort;             // global vector to store the toposort in reverse order

void dfs2(int u) {    // change function name to differentiate with original dfs
  dfs_num[u] = DFS_BLACK;
  for (auto &v : AL[u])
    if (dfs_num[v.first] == DFS_WHITE)
      dfs2(v.first);
  topoSort.push_back(u); }                   // that is, this is the only change

#define DFS_GRAY 2              // one more color for graph edges property check
vi dfs_parent;      // to differentiate real back edge versus bidirectional edge

void graphCheck(int u) {               // DFS for checking graph edge properties
  dfs_num[u] = DFS_GRAY;   // color this as DFS_GRAY (temp) instead of DFS_BLACK
  for (auto &v : AL[u]) {
    if (dfs_num[v.first] == DFS_WHITE) {     // Tree Edge, DFS_GRAY to DFS_WHITE
      dfs_parent[v.first] = u;                  // parent of this children is me
      graphCheck(v.first);
    }
    else if (dfs_num[v.first] == DFS_GRAY) {             // DFS_GRAY to DFS_GRAY
      if (v.first == dfs_parent[u])          // to differentiate these two cases
        printf(" Bidirectional (%d, %d) - (%d, %d)\n", u, v.first, v.first, u);
      else  // the most frequent application: check if the given graph is cyclic
        printf(" Back Edge (%d, %d) (Cycle)\n", u, v.first);
    }
    else if (dfs_num[v.first] == DFS_BLACK)             // DFS_GRAY to DFS_BLACK
      printf(" Forward/Cross Edge (%d, %d)\n", u, v.first);
  }
  dfs_num[u] = DFS_BLACK;     // after recursion, color this as DFS_BLACK (DONE)
}

vi dfs_low;       // additional information for articulation points/bridges/SCCs
vi articulation_vertex;
int dfsNumberCounter, dfsRoot, rootChildren;

void articulationPointAndBridge(int u) {
  dfs_low[u] = dfs_num[u] = dfsNumberCounter++;      // dfs_low[u] <= dfs_num[u]
  for (auto &v : AL[u]) {
    if (dfs_num[v.first] == DFS_WHITE) {                          // a tree edge
      dfs_parent[v.first] = u;
      if (u == dfsRoot) rootChildren++;  // special case, count children of root

      articulationPointAndBridge(v.first);

      if (dfs_low[v.first] >= dfs_num[u])              // for articulation point
        articulation_vertex[u] = true;           // store this information first
      if (dfs_low[v.first] > dfs_num[u])                           // for bridge
        printf(" Edge (%d, %d) is a bridge\n", u, v.first);
      dfs_low[u] = min(dfs_low[u], dfs_low[v.first]);       // update dfs_low[u]
    }
    else if (v.first != dfs_parent[u])       // a back edge and not direct cycle
      dfs_low[u] = min(dfs_low[u], dfs_num[v.first]);       // update dfs_low[u]
} }

vi S, visited;                                    // additional global variables
int numSCC;

void tarjanSCC(int u) {
  dfs_low[u] = dfs_num[u] = dfsNumberCounter++;      // dfs_low[u] <= dfs_num[u]
  S.push_back(u);           // stores u in a vector based on order of visitation
  visited[u] = 1;
  for (auto &v : AL[u]) {
    if (dfs_num[v.first] == DFS_WHITE)
      tarjanSCC(v.first);
    if (visited[v.first])                                // condition for update
      dfs_low[u] = min(dfs_low[u], dfs_low[v.first]);
  }

  if (dfs_low[u] == dfs_num[u]) {         // if this is a root (start) of an SCC
    printf("SCC %d:", ++numSCC);            // this part is done after recursion
    while (1) {
      int v = S.back(); S.pop_back(); visited[v] = 0;
      printf(" %d", v);
      if (u == v) break;
    }
    printf("\n");
} }

int main() {
  /*
  // Use the following input:
  // Graph in Figure 4.1
  9
  1 1 0
  3 0 0 2 0 3 0
  2 1 0 3 0
  3 1 0 2 0 4 0
  1 3 0
  0
  2 7 0 8 0
  1 6 0
  1 6 0

  // Example of directed acyclic graph in Figure 4.4 (for toposort)
  8
  2 1 0 2 0
  2 2 0 3 0
  2 3 0 5 0
  1 4 0
  0
  0
  0
  1 6 0

  // Example of directed graph with back edges
  3
  1 1 0
  1 2 0
  1 0 0

  // Left graph in Figure 4.6/4.7/4.8
  6
  1 1 0
  3 0 0 2 0 4 0
  1 1 0
  1 4 0
  3 1 0 3 0 5 0
  1 4 0

  // Right graph in Figure 4.6/4.7/4.8
  6
  1 1 0
  5 0 0 2 0 3 0 4 0 5 0
  1 1 0
  1 1 0
  2 1 0 5 0
  2 1 0 4 0

  // Directed graph in Figure 4.9
  8
  1 1 0
  1 3 0
  1 1 0
  2 2 0 4 0
  1 5 0
  1 7 0
  1 4 0
  1 6 0
  */

  freopen("01_in.txt", "r", stdin);

  int V; scanf("%d", &V);
  AL.assign(V, vii()); // assign blank vectors of pair<int, int>s to Al
  for (int i = 0; i < V; i++) {
    int total_neighbors; scanf("%d", &total_neighbors);
    for (int j = 0; j < total_neighbors; j++) {
      int id, weight; scanf("%d %d", &id, &weight);
      AL[i].emplace_back(id, weight);
    }
  }

  printThis("Standard DFS Demo (the input graph must be UNDIRECTED)");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE);    // this sets all vertices' state to DFS_WHITE
  for (int i = 0; i < V; i++)                   // for each vertex i in [0..V-1]
    if (dfs_num[i] == DFS_WHITE)            // if that vertex is not visited yet
      printf("CC %d:", ++numCC), dfs(i), printf("\n");          // 3 lines here!
  printf("There are %d connected components\n", numCC);

  printThis("Flood Fill Demo (the input graph must be UNDIRECTED)");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE);
  for (int i = 0; i < V; i++)
    if (dfs_num[i] == DFS_WHITE)
      floodfill(i, ++numCC);
  for (int i = 0; i < V; i++)
    printf("Vertex %d has color %d\n", i, dfs_num[i]);

  // make sure that the given graph is DAG
  printThis("Topological Sort (the input graph must be DAG)");
  topoSort.clear();
  dfs_num.assign(V, DFS_WHITE);
  for (int i = 0; i < V; i++)            // this part is the same as finding CCs
    if (dfs_num[i] == DFS_WHITE)
      dfs2(i);
  reverse(topoSort.begin(), topoSort.end());                 // reverse topoSort
  for (int i = 0; i < (int)topoSort.size(); i++)       // or you can simply read
    printf(" %d", topoSort[i]);           // the content of `topoSort' backwards
  printf("\n");

  printThis("Graph Edges Property Check");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE); dfs_parent.assign(V, -1);
  for (int i = 0; i < V; i++)
    if (dfs_num[i] == DFS_WHITE)
      printf("Component %d:\n", ++numCC), graphCheck(i);       // 2 lines in one

  printThis("Articulation Points & Bridges (the input graph must be UNDIRECTED)");
  dfsNumberCounter = 0; dfs_num.assign(V, DFS_WHITE); dfs_low.assign(V, 0);
  dfs_parent.assign(V, -1); articulation_vertex.assign(V, 0);
  printf("Bridges:\n");
  for (int i = 0; i < V; i++)
    if (dfs_num[i] == DFS_WHITE) {
      dfsRoot = i; rootChildren = 0;
      articulationPointAndBridge(i);
      articulation_vertex[dfsRoot] = (rootChildren > 1); }       // special case
  printf("Articulation Points:\n");
  for (int i = 0; i < V; i++)
    if (articulation_vertex[i])
      printf(" Vertex %d\n", i);

  printThis("Strongly Connected Components (the input graph must be DIRECTED)");
  dfs_num.assign(V, DFS_WHITE); dfs_low.assign(V, 0); visited.assign(V, 0);
  dfsNumberCounter = numSCC = 0;
  for (int i = 0; i < V; i++)
    if (dfs_num[i] == DFS_WHITE)
      tarjanSCC(i);

  return 0;
}
